File: android/app/src/main/java/com/vpn/android/vpn/state/FailureReason.java
package com.vpn.android.vpn.state;

import com.vpn.android.api.ApiException;

import java.io.IOException;
import java.util.Locale;

/**
 * Why a connection attempt ended in ERROR, so the screen can say something
 * more useful than "Error" and offer the one action that helps. Pure Java,
 * like the rest of this package, so it is covered by plain JUnit.
 */
public enum FailureReason {
    /** No active plan — expired, out of traffic, or never had one. Fix: get a plan. */
    NO_SUBSCRIPTION,
    /** The sign-in is no longer valid and could not be renewed. Fix: sign in. */
    SESSION_EXPIRED,
    /** The account has no server it can use right now. Fix: try later / another region. */
    NO_SERVERS,
    /** Our API could not be reached at all. Fix: check the connection, retry. */
    NETWORK,
    /** Anything else. */
    UNKNOWN;

    public static FailureReason classify(Throwable error) {
        if (error instanceof ApiException) {
            ApiException api = (ApiException) error;
            if (api.httpCode == 401) return SESSION_EXPIRED;
            String message = api.getMessage() == null ? "" : api.getMessage().toLowerCase(Locale.ROOT);
            if (message.contains("subscription")) return NO_SUBSCRIPTION;
            return UNKNOWN;
        }
        if (error instanceof IOException) return NETWORK;
        String message = error == null || error.getMessage() == null
                ? "" : error.getMessage().toLowerCase(Locale.ROOT);
        if (message.contains("no subscription links") || message.contains("no usable subscription links")) {
            return NO_SERVERS;
        }
        return UNKNOWN;
    }
}

File: desktop/src/shared/failureReason.ts
/**
 * Why a connection attempt ended in ERROR, so the UI can say something more
 * useful than "Error" and offer the one action that helps. Pure — same
 * classification as the Android client's FailureReason.
 */
export type FailureReason = 'NO_SUBSCRIPTION' | 'SESSION_EXPIRED' | 'NO_SERVERS' | 'NETWORK' | 'UNKNOWN';

export function classifyFailure(error: unknown): FailureReason {
  const httpCode =
    error && typeof error === 'object' && 'httpCode' in error ? (error as { httpCode: unknown }).httpCode : undefined;
  const message = error instanceof Error ? error.message.toLowerCase() : '';
  if (typeof httpCode === 'number') {
    if (httpCode === 401) return 'SESSION_EXPIRED';
    if (message.includes('subscription')) return 'NO_SUBSCRIPTION';
    return 'UNKNOWN';
  }
  if (message.includes('no subscription links') || message.includes('no usable subscription links')) {
    return 'NO_SERVERS';
  }
  // Node's fetch reports every connectivity failure as a TypeError ("fetch failed").
  if (error instanceof TypeError || message.includes('fetch failed') || message.includes('network')) {
    return 'NETWORK';
  }
  return 'UNKNOWN';
}
